ez-launch/sources/ez_launch.move

```diff
@@ -0,0 +1,413 @@
+        let object_signer_address = signer::address_of(&object_signer);
+        assert!(exists<CollectionDetails>(object_signer_address), error::invalid_state(ECOLLECTION_DETAILS_DOES_NOT_EXIST));
+        let collection_details_obj = borrow_global_mut<CollectionDetails>(object_signer_address);
+        let to_be_minted = collection_details_obj.to_be_minted;
+        let length = vector::length(&to_be_minted);
+
+        assert!(length > 0, error::permission_denied(EMINTING_HAS_ENDED));
+        assert!(collection_details_obj.ready_to_mint, error::permission_denied(EMINTING_HAS_NOT_STARTED_YET));
+
+        // (jill) check against whitelist
+
+        vector::for_each_ref(&collection_details_obj.coin_payments, |coin_payment| {
+            let coin_payment: &CoinPayment<AptosCoin> = coin_payment;
+            coin_payment::execute(user, coin_payment);
+        });
+
+        
+        let index = if (collection_details_obj.random_mint) {
+            timestamp::now_seconds() % length
+        } else {
+            length - 1
+        };
+        let token = *vector::borrow(&to_be_minted, index);
+        let token_address = object::object_address(&token);
+        let user_address = signer::address_of(user);
+        object::transfer(&object_signer, token, user_address);
+        vector::pop_back(&mut to_be_minted);
+
+        if (collection_details_obj.is_soulbound) {
+            assert!(token_components::contains_token_refs(token_address), ETOKEN_REFS_DOES_NOT_EXIST);
+            token_components::disable_ungated_transfer<Token>(&object_signer, token);
+        };
+        token
+    }
+
+    fun assert_owner<T: key>(owner: address, object: Object<T>) {
+        assert!(object::owner(object) == owner, error::permission_denied(ENOT_OWNER));
+    }
+
+    fun authorized_collection_signer(creator: &signer, collection: Object<Collection>): signer acquires EZLaunchRefs {
+        let refs = authorized_borrow(creator, collection);
+        object::generate_signer_for_extending(&refs.extend_ref)
+    }
+
+    fun collection_signer(collection: Object<Collection>): signer acquires EZLaunchRefs {
+        let refs = borrow(collection);
+        object::generate_signer_for_extending(&refs.extend_ref)
+    }
+
+    inline fun authorized_borrow(creator: &signer, collection: Object<Collection>): &EZLaunchRefs acquires EZLaunchRefs {
+        assert_owner<Collection>(signer::address_of(creator), collection);
+        borrow(collection)
+    }
+
+    inline fun borrow(collection: Object<Collection>): &EZLaunchRefs acquires EZLaunchRefs {
+        let collection_address = object::object_address(&collection);
+        assert!(
+            exists<EZLaunchRefs>(collection_address),
+            error::not_found(EEZ_LAUNCH_REFS_DOES_NOT_EXIST)
+        );
+        borrow_global<EZLaunchRefs>(collection_address)
+    }
+
+    // ================================= Tests  ================================= //
+
+    #[test_only]
```

Comment:
Let's have tests in a separate directory under `tests`. We shouldn't be writing tests in the same file now, as we now can write tests in a `tests` directory. 

This also makes it easier to read and reason about the module.

Reply:
we're having a lot of entry functions here which don't return objects at all, so we have couple internal funcs that are does return object for us to test against. thats why they are in the same file as i don't want to expose internal funcs. any better suggestions on that?

Reply:
Internal functions that return objects can simply be public functions, and you call the public functions from the test file :)